File: src/cxx-dev/LiENa/lienaRealTime/lienaDiagnosisTask.cpp
#include "lienaDiagnosisTask.h"


/**
 *
 * @brief lienaDiagnosisTask::lienaDiagnosisTask
 *
 */
lienaDiagnosisTask::lienaDiagnosisTask(){
    this->rtFlag = false;
    this->connexionFailed = false;
    this->rtPeriod = 2;
    this->heartBeatMessageCount = 0;
}

//!------------------------------------------------------------------------------
//!
//!
//! \brief lienaDiagnosisTask::~lienaDiagnosisTask
//!
lienaDiagnosisTask::~lienaDiagnosisTask(){

}

//!------------------------------------------------------------------------------
//!
//! \brief lienaDiagnosisTask::append
//! \param msg
//!
void lienaDiagnosisTask::append(lienaHeartBeatMessage* msg){
    this->heartBeatMessageCount += 1;
    mutex.lock();
    this->heartBeatMessage.append(msg);
    mutex.unlock();

    if(this->heartBeatMessageCount == 1){
        qDebug()<<"diagnosis task launch";
        this->launch();
    }
}

//!------------------------------------------------------------------------------
//!
//! \brief lienaDiagnosisTask::launch
//!
void lienaDiagnosisTask::launch(){
    this->rtFlag = true;
    this->connectionStatus = 1;
    this->start();
}

//!//!------------------------------------------------------------------------------
//! //! \brief lienaDiagnosisTask::get_sequence_length
//! \return
//!
int lienaDiagnosisTask::get_sequence_length(){
    int ret;
    mutex.lock();
    ret = this->heartBeatMessage.size();
    mutex.unlock();
    return ret;
}

//!------------------------------------------------------------------------------
//!
//! \brief lienaDiagnosisTask::get_latest_message
//! \return
//!
lienaHeartBeatMessage *lienaDiagnosisTask::get_latest_message(){
    lienaHeartBeatMessage *ret = nullptr;
    mutex.lock();
    ret = this->heartBeatMessage.at(0);
    mutex.unlock();
    return ret;
}

//!------------------------------------------------------------------------------
//!
//! \brief lienaDiagnosisTask::deleteFrontDatagram
//!
void lienaDiagnosisTask::deleteFrontDatagram(){
    mutex.lock();
    this->heartBeatMessage.pop_front();
    mutex.unlock();
}

//!------------------------------------------------------------------------------
//!
//! \brief lienaDiagnosisTask::connexionFailedRecovered
//!
void lienaDiagnosisTask::connexionFailedRecovered(){
    this->connexionFailed = false;
}

//!------------------------------------------------------------------------------
//!
//! \brief lienaDiagnosisTask::run
//!
void lienaDiagnosisTask::run(){
    while(this->rtFlag){

        if(this->connexionFailed){
            qDebug()<<"lienaDiagnosisTask | wait for network failure handling";
            sleep(1);
            continue;
        }

        int len = this->get_sequence_length();
        if(len > 0){
            this->connectionStatus = 1;
            //lienaHeartBeatMessage* msg = this->get_latest_message();
            this->deleteFrontDatagram();
        }
        else if(len == 0){
            this->connectionStatus *= 2;
            if(this->connectionStatus > 8){
                qDebug()<<"lenaDiagnosisTask | error";
                emit lostConnection();
                this->connectionStatus = 1;
                connexionFailed = true;
            }
        }
        sleep(this->rtPeriod);
    }
}

File: src/pyDev/LiENa/LiENaRealTimeTask/LienaDiagnosisTask.py
import threading
import time
from PyQt5.QtCore import QObject, pyqtSignal
from LiENa.LiENaBasic.lienaDefinition import *


class LienaDiagnosisTask(QObject):

    lostConnexion = pyqtSignal()

    def __init__(self):
        super(LienaDiagnosisTask, self).__init__()
        self.heartBeatMessageListLock = threading.Lock()
        self.heartBeatMessageList = []
        self.flag = False
        self.heartBeatMessageArrived = False

        self.connection_failed = False

        self.period = 2
        self.connexionStatus = 1
        self.heartBeatMessageCount = 0

    def heartbeat_message_arrived(self):
        return self.heartBeatMessageArrived

    def set_period(self, period):
        self.period = period

    def terminate(self):
        self.flag = False

    def launch(self):
        self.flag = True
        self.heartBeatMessageList = []
        self.connexionStatus = 1
        task = threading.Thread(None, self.diagnosis)
        task.start()

    def get_sequence_length(self):
        self.heartBeatMessageListLock.acquire()
        ret = len(self.heartBeatMessageList)
        self.heartBeatMessageListLock.release()
        return ret

    def get_latest_message(self):
        ret = None
        self.heartBeatMessageListLock.acquire()
        ret = self.heartBeatMessageList.pop(0)
        self.heartBeatMessageListLock.release()
        return ret

    def append(self, msg):
        self.heartBeatMessageCount += 1
        self.heartBeatMessageListLock.acquire()
        self.heartBeatMessageList.append(msg)
        self.heartBeatMessageListLock.release()

        if self.heartBeatMessageCount == 1:
            if DEBUG:
                print("diagnosis task launched")
            self.launch()

    def connection_failed_recovery(self):
        self.connection_failed = False

    def diagnosis(self):
        while self.flag:

            if self.connection_failed:
                print("LienaDiagnosisTask | wait for network failure handling ")
                time.sleep(1)
                continue

            if self.get_sequence_length() > 0:
                self.connexionStatus = 1
                msg = self.get_latest_message()
            elif self.get_sequence_length() == 0:
                self.connexionStatus *= 2
                if self.connexionStatus > 8:
                    self.lostConnexion.emit()
                    self.connexionStatus = 1
                    self.connection_failed = True

            time.sleep(self.period)
